model/manifest.go

```diff
@@ -60,44 +65,50 @@ type CVOptions struct {
 	RoleName      string   `json:"role_name,omitempty" yaml:"role_name,omitempty"`
 }
 
-// Internal structure for extracting our CVOptions into a struct
-type internalConfigurationVariable struct {
-	CVOptions CVOptions `yaml:"options"`
-}
-
-// Since we want all keys below options: but still need access to a number of fissile special
-// options.
-type internalVariableDefinitions struct {
-	Variables []*internalConfigurationVariable `yaml:"variables"`
-}
-
 // CertParams was copied from config-server/types/certificate_generator.go for on the fly parsing of certificate options
 type CertParams struct {
-	CommonName       string   `yaml:"common_name"`
 	AlternativeNames []string `yaml:"alternative_names" json:"subject_names,omitempty"`
 	IsCA             bool     `yaml:"is_ca"`
+	AppendKubeCA     bool     `yaml:"append_kube_ca"`
 	CAName           string   `yaml:"ca"`
 	ExtKeyUsage      []string `yaml:"extended_key_usage"`
 }
 
 // GetManifest loads a manifest from file or string
-func GetManifest(r io.Reader) (Manifest, error) {
+func GetManifest(r io.Reader, env map[string]string) (Manifest, error) {
 	var manifest Manifest
 
 	data, err := ioutil.ReadAll(r)
 	if err != nil {
 		return manifest, err
 	}
 
-	err = yaml.Unmarshal(data, &manifest)
+	// Read raw (no schema) manifest so we can expand templates in the variables section.
+	var raw map[string]interface{}
+	err = yaml.Unmarshal(data, &raw)
+	if err != nil {
+		return manifest, err
+	}
+	if raw["variables"] == nil {
+		return manifest, errors.New("'variables' section not found in manifest")
+	}
+
+	_, err = expandTemplates(raw["variables"], env)
```

Comment:
> it's not modifying raw["variables"] in place is it? It might be better to just copy in that case…

It is doing just that. Recursively copying larger and larger tree structures in general feels like a bad idea, although in this particular instance the structures are reasonably small.

I did run into some problems when trying to implement everything via copying, but I can't remember anymore (this commit is already 6 weeks old).

I'm happy to add a comment, but not really keen to rewrite this from scratch, given that it works "as documented" (`expandTemplates` docs says that it expands "in-place").

Reply:
Fixed in 8cf5338

Reply:
I've documented what we are doing, so resolving this one as well. :)